### ISSUE TITLE
Invalid Patch Content order in Firefox

### ISSUE THREAD

**Issue by andwah** [PuppetJs/PuppetJs#29]
The content order of an patch differ between Firefox and Chrome

***Javascript code snippet***
```
this.templateInstance.model.AuthToken$ = token;
this.templateInstance.model.SignIn$ = null;  // This is an operation handler
```

***Chrome:*** (Correct order)
```
[
    {
        "op": "replace",
        "path": "/menu/SignInApp/AuthToken$",
        "value": "m8VuWg_7coG8ikZdlgLuCA2"
    },
    {
        "op": "replace",
        "path": "/menu/SignInApp/SignIn$",
        "value": null
    }
]
```

***FireFox:*** (Incorrect order)
```
[
    {
        "op": "replace",
        "path": "/menu/SignInApp/SignIn$",
        "value": null
    },
    {
        "op": "replace",
        "path": "/menu/SignInApp/AuthToken$",
        "value": "SkVDaOCX2zxW0J0NQB081A2"
    }
]
```

**Comment by andwah** [comment]
:+1: i think the order is important.  Imagine you change the "Name" property then invoke the "Save" operation.. then you may not get the right "Name"

**Comment by warpech** [comment]
@mattiasmak has reported a similar problem with his app few days ago. 

Let's reopen this issue to collect some thoughts. 

The only idea that I have is to do something like:

```js
var that = this;
this.templateInstance.model.AuthToken$ = token;
this.templateInstance.model.generate(); //triggers JSON Patch generate
setTimeout(function() {
  that.templateInstance.model.SignIn$++;
}, 1);
```

But how to do that in HTML?

```html
<button value="{{model.SignIn$::click}}" model="{{model}}" onclick="model.generate(); ++this.value">Add carrots</button>
```

Or, using a static method:

```html
<button value="{{model.SignIn$::click}}" onclick="Puppet.generate(); ++this.value">Add carrots</button>
```

**Comment by mattiasmak** [comment]
Is a helper method in puppet a solution? Send all properties to the methon and let puppet create the patch in the correct order? 

puppet.triggerpatch({Somedata:1, Someotherdata:2, Evenmoredata:'data'});

Or do you want puppet.js to be as clean as possible?

**Comment by mattiasmak** [comment]
Or we just use a | seperated string for now... But I don't like it.

**Comment by mtornwall** [comment]
Darn, this just bit me too. :-)

While it's unfortunate that we cannot guarantee that the program write order is reflected in the patch, it isn't an insurmountable obstacle.

What we need is something akin to a [memory barrier](https://en.wikipedia.org/wiki/Memory_barrier) - a way of forcing Puppet to make sure that all writes that occurred before the barrier have been applied  (notice the vague wording) after the barrier:

    // Modify some properties that require no particular ordering.
    model.Property1$ = "foo";
    model.Property2$ = "bar";

    // Writes to Property1$ and Property2$ may or may not have
    // taken place yet, and if so they can be in any order.
    Puppet.barrier();

    // Writes to Property1$ and Property2$ will definitely have taken place now.
    model.Submit$++;

Without being familiar with the specifics of the PuppetJS implementation, it seems to me that a simple and correct implementation strategy would be for the barrier to force a dirty check, then generate and enqueue (or send) a patch. Perhaps this was what you were getting at @warpech? (Though I don't think this should require a `setTimeout`)

Writes following the barrier would be recorded in a new patch, which (if this makes sense from an optimization standpoint) could still be bundled in the same message as the pre-barrier patch.

**Comment by tomalec** [comment]
I'm afraid that as `Object.observe` was deprecated, it may hit us more often :(

I really like the @mtornwall Barier approach. So we could have methods to:
 - `.touch` - just calls dirty-check, to gather all already made changes, and put proceeding JSON Patch operation objects after
 - `.sync` - `.touch` and sync with server-side
 - `.async` - Similar to Polymer, just a sugar for @warpech `setTimeout` that will also make sure that already made patches were douched before callback execution
(method names to be decided)

The question is whether we should make it:
 1. instance `puppet.method()` - but then it would be hard to use it in HTML and with bindings to nested nodes
 2. static `Puppet.method()` - but should it the execute it for all instances, all connections, and all models? - performance hurts, asking for trouble with other debugging unpredictable changes
3. `Puppet.method(model)` - easier than first, cheaper that second,
4. `Puppet.method(node_somewhere_in_model)` - easier than first, cheaper that second, traversing tree may cost.

**Comment by mtornwall** [comment]
@tomalec Great idea with the `touch`/`sync`/`async` function set; I think it's preferable to just the barrier I proposed above.

Regarding static vs instance I really prefer instance (despite how it appears in my example :-)). Perhaps Puppet could augment the model object with functions to do this? Alternatively, the model could have a property through which the Puppet instance could be accessed?

~~~
model.__puppet.touch();
~~~

**Comment by tomalec** [comment]
I do not like adding too many properties and methods, not to polute users
space, and not to have too much about GC, there is already parent property,
so in worst case we could add such property on root level.
12 sty 2016 14:28 "Martin Törnwall" <notifications@github.com> napisał(a):

**Comment by mtornwall** [comment]
@tomalec Yeah, you're right. Perhaps `Puppet.method(model)` is the best choice, then? It's pretty easy to use, doesn't pollute the model object and has clear semantics. :+1:

**Comment by mtornwall** [comment]
Any update on this? Lately I've been getting diffuse reports suggesting that this issue occurs more frequently than expected.

**Comment by warpech** [comment]
@mtornwall No, still to separate batches of patches you should better trigger a `mouseup` event, as described here: https://github.com/PuppetJs/PuppetJs#generating-patches-based-on-local-changes

**Comment by mtornwall** [comment]
@warpech The problem with this approach is that it is still asynchronous. I don't really know when the `mouseup` event will be processed. By contrast, the barrier approach I proposed would be synchronous and therefore guarantee a partial write ordering across the barrier.

**Comment by miyconst** [comment]
Just now me and @ubbeK were hit by this issue. A client side patch was sent to server too late, by that time the modified object was removed from the view-model and server fails.

So, I +1 this issue.

The scenario we have is:

- User clicks a button.
- The `onclick` listener executed.
- Local view-model changed.
- PuppetJs observes and sends changes.
- Server responses with another patch.
- Polymer observer executed.
- The observer changes local view-model.
- **PuppetJs waits for  a user action.**

**Comment by malx122** [comment]
We have been hit a LOT by this latetly. Everytime there is two or more changes to the model they are sent in the REVERSE order of how they are found, kind of not so intuitive. Normally when I design a JSON on the serverside I finish with the "buttons" executing everything. But lets say that I have a text that should be sent on ok. Then the text (that triggers on lose focus) are sent after the ok-button if not the user click somewhere else first.

**Comment by warpech** [comment]
Today there was a similar problem in Starcounter: https://github.com/Starcounter/Starcounter/issues/4044

**Comment by dan31** [comment]
Looks to be a decently old issue. @tomalec, what do you think about this one in 2017?

**Comment by warpech** [comment]
I have added it to goals 2017 in September/October (as PuppetJs maintainance/fixes). This is very important issue, but others are more prioritised.

**Comment by alshakero** [comment]
@warpech isn't thing going to be solved when we switch to proxies?

**Comment by warpech** [comment]
You are right, the proxy solves the problem of observing the changes in wrong order, which is what the OP says in https://github.com/PuppetJs/PuppetJs/issues/29#issue-41708810

But there is also another problem. Consider the following code:

```js
puppet.obj.FirstName$ = "Jim";
puppet.obj.LastName$ = "Carrey";
puppet.obj.Save$ = 1 + puppet.obj.Save$;
```

This will be detected by the proxy as 3 distinct changes, right?

Will the PuppetJs send this as 3 separate sequences with a single patch, or as 1 sequence with three patches? I guess the former is easier to achieve, but actually there must be a way to also enforce the latter.

**Comment by alshakero** [comment]
@warpech I see. I guess it's easily doable. We can delay sending without losing any data.

**Comment by warpech** [comment]
I think that good ideas are here: https://github.com/PuppetJs/PuppetJs/issues/29#issuecomment-170904351. But I agree, let's get back to this after proxies.

**Comment by tomalec** [comment]
https://github.com/Palindrom/Palindrom/issues/128#issuecomment-291535681 shows that we need it at the same time we have release proxies. 

I thought about something closer to what we have on server-side:
```js
var node = root.some.path.deeper;
palindrom.batch/transaction(function(root){
    for(i = 0; i < 100; i++){
        node.arr[i] = 1;
    }
});
```

**Comment by warpech** [comment]
I am afraid that if the data change is performed asynchronously by some framework, we might be not possible to wrap changes in a function. Maybe we also should provide a flattened version - `palindrom.batchStart`, `palindrom.batchEnd`.

Also, how do you obtain the `palindrom` instance? Do you get there reference from `juicy-html`?

**Comment by alshakero** [comment]
@warpech that's was my preliminary idea. I also have another idea that can be the best but might force me re-do the tests 😢 

Since until today, we assume that jsonpatch is async, we can leverage this assumption and make proxies async too. 

Instead of calling the callback instantly when a change occurs. We do this:

```js
 // => a change happened 
patches.push(patch)
globalTimeout && clearTimeout(globalTimeout);
globalTimeout = setTimeout(emitPatches(patches), 1)
```

And by this, we wait for any blocking changes to happen and then emit.

**Comment by tomalec** [comment]
If data changes are done via automagic framework, it will behave as usual - send patch for each change.
If developer would like to do some advanced binding he/she would have to do it manually.
As such batched changes may also apply to data-binding framework, and each framework may support (or not) it in different manner.

> Also, how do you obtain the palindrom instance? Do you get the reference from `juicy-html`?

Via `document.querySelector('puppet-client');`

**Comment by tomalec** [comment]
> Since until today we assume that jsonpatch is async, we can leverage this assumption and make proxies async too.

I would really like to get away of unnecessarily complicated flows.

**Comment by warpech** [comment]
> Since until today we assume that jsonpatch is async, we can leverage this assumption and make proxies async too.

I was thinking about this kind of throttling too, but let's try to avoid this unless forced to. One thing that I will dislike for sure - I don't want Palindrom to be sometimes sync, sometimes async. It should be consistent in that manner.

**Comment by warpech** [comment]
Putting this to freezer because as of Palindrom 3.0.0, patches are generated as real time, not with dirty checking. Patch grouping should be less needed now.

**Comment by warpech** [comment]
Related: https://github.com/Palindrom/JSONPatcherProxy/issues/7#issuecomment-334085871

**Comment by mtornwall** [comment]
> Putting this to freezer because as of Palindrom 3.0.0, patches are generated as real time, not with dirty checking. Patch grouping should be less needed now.

Patch grouping maybe, but there's presumably still the problem of ordering within a patch? For example if I write properties `a`, `b`, `c`, in that order, am I guaranteed that the order of operations within the patch will match the program order of the writes? If not, the "fence" solution I suggested above is still both relevant and required.
